Resume shipment from canceled logic

urbanautomaton opened the pull request:
Hi,

While documenting the `Spree::Shipment` state machine today I noticed some bugs in the logic handling resumption from a canceled state. Specifically, there are duplicate, contradictory transitions from the `:canceled` state on a `:resume` event:

![spree shipment_state](https://cloud.githubusercontent.com/assets/383376/8569859/a330831a-2573-11e5-9418-8242bbb88e9c.jpg)

The [relevant state machine code](https://github.com/spree/spree/blob/10df25f1fbd852eef0491d7ffa3331fe1b4b9042/core/app/models/spree/shipment.rb#L64-L70) looks like this:

``` ruby
# core/app/models/spree/shipment.rb
event :resume do
  transition from: :canceled, to: :ready, if: lambda { |shipment|
    shipment.determine_state(shipment.order) == :ready
  }
  transition from: :canceled, to: :pending, if: lambda { |shipment|
    shipment.determine_state(shipment.order) == :ready
  }
  transition from: :canceled, to: :pending
end
```

This is broken in a couple of ways:
- `Spree::Shipment#determine_state` returns strings, not symbols, so the comparisons will always fail
- Even if the comparisons are fixed, the second condition never applies:
  - if the order state is `'ready'` then the first condition applies, and the shipment will transition to `:ready`.
  - If the order state is not `'ready'`, then the second condition is overridden by the final, duplicate clause, and the shipment transitions to `:pending` regardless.

The current effect is that the shipment always transitions from `:canceled` to `:pending` on a `:resume` event.
### Tests

There are two passing tests for this behaviour, but they're a bit broken themselves.

The first, [asserting that a shipment with a `'ready'` order will transition to `:ready`](https://github.com/spree/spree/blob/8bd0f1ce1cf2ca7185ed18e65c596f85e70a7131/core/spec/models/spree/shipment_spec.rb#L442-L450), passes only because the `#determine_state` method has been [incorrectly stubbed](https://github.com/spree/spree/blob/8bd0f1ce1cf2ca7185ed18e65c596f85e70a7131/core/spec/models/spree/shipment_spec.rb#L446) to return a symbol, not a string.

The second, contradictory test, [asserting that a shipment with a `'ready'` order will transition to `:pending`](https://github.com/spree/spree/blob/8bd0f1ce1cf2ca7185ed18e65c596f85e70a7131/core/spec/models/spree/shipment_spec.rb#L459-L467), passes only because of the third, unconditional state machine transition (and because it _correctly_ stubs the `#determine_state` method to return a string, meaning the first conditional doesn't succeed).
### Fixes

This PR currently only has two commits - one to fix the incorrect stubbing in the tests, and another to fix the comparisons so that they compare strings. The result of these changes is that the second test now correctly fails.

I would be happy to provide a full fix, but I'm not sure what the expected behaviour is. The transition logic was introduced in [this commit](https://github.com/spree/spree/commit/82ba2a7828a9572398637f95b4a9a91fae050739), which doesn't provide much clue as to the intentions behind the change.

As best I can tell the current behaviour has been in place since [early 2013](https://github.com/spree/spree/commit/f156409cf285e8a0bfd5e77e6df559b70a70e245#commitcomment-3302976). It _seems_ as though the original intention was to transition to `:ready` if `#determine_state` returned `'ready'`, and `:pending` otherwise, but it'd be great if someone with a better idea of the implications of that change could weigh in before I start making changes. If the bug has gone unnoticed for ~2 years, maybe that complexity is unnecessary after all.

Cheers,
Simon

urbanautomaton committed:
Correct stub to return string, per the method it stubs

The comparisons made in the conditionals for the resume event are faulty,
comparing the results of `Shipment#determine_state` with symbols, when that
method returns only strings. The test only passed because the method was
stubbed to return a symbol.

Correct the stub to make the test fail properly.

urbanautomaton committed:
Fix shipment :resume event conditional comparisons

`Spree::Shipment#determine_state returns strings, not symbols.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://urbanautomaton-hotfix-resume-shipment-from-canceled.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://urbanautomaton-hotfix-resume-shipment-from-canceled.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

JDutil commented:
I believe your right about the intended behavior, but not sure how it's gone this long without anyone noticing it doesn't work.

urbanautomaton commented:
The (probably) intended behaviour was added in 82ba2a78 and broken in f156409c - there are no release tags [in that range](https://github.com/spree/spree/compare/82ba2a78%5E...f156409c), so it seems likely that it was never released, and from every user's perspective, the original behaviour never changed. This could explain why the dead code never got picked up.

If I'm to reinstate that behaviour, then, it's really equivalent to adding it for the first time. Perhaps @cmar can help us with the reasons it was added in the first place?

urbanautomaton commented:
Okay, in the absence of any other clues, and given that the conditional transitions appear to have never been released in a working state, my preference would be to completely remove the conditionals. That way the code says exactly what it currently does, i.e. always transition from `:cancel` to `:pending` on `:resume`.

Any objections to that approach?

JDutil commented:
@urbanautomaton yea please update to change to pending if determine state returns pending.

urbanautomaton commented:
Hi @JDutil - sorry, I think there's some confusion. I'm proposing to change the code so the only allowed transition is from `:canceled` to `:pending`, regardless of what `#determine_state` returns:

```ruby
# core/app/models/spree/shipment.rb
event :resume do
  transition from: :canceled, to: :pending
end
```

That's the real current behaviour (due to the symbol/string comparison bug), and as best I can tell is the only behaviour that has ever been in a released version of spree.

We could do this instead:

```ruby
# core/app/models/spree/shipment.rb
event :resume do
  transition from: :canceled, to: :ready, if: lambda { |shipment|
    shipment.determine_state(shipment.order) == 'ready'
  }
  transition from: :canceled, to: :pending
end
```

which I think we agree was the original intention, but to my mind this is a breaking change at this point, and I don't know what the implications are for client code, so I'm pretty wary of this choice.

JDutil commented:
:+1:

JDutil commented:
I think we should fix it to the intended behavior on master branch, and can leave stable branches as they are.

urbanautomaton committed:
Correct Spree::Shipment#resume specs and remove redundant statement

urbanautomaton commented:
Okay, I've updated the tests to reflect what we think should be the case, and removed a redundant clause from the state machine.

One more thing: the `#determine_state` method returns one of `%w[ ready pending shipped ]`. Should the shipment transition to `:shipped` on `#resume` when `#determine_state` returns `'shipped'`? Currently it will transition to `:pending`.
